File: daq/slc/dqm/src/CharArray.cc
#include "daq/slc/dqm/CharArray.h"

#include <sstream>

using namespace Belle2;

CharArray::CharArray() throw() : NumberArray(0), _value_v(NULL) {}

CharArray::CharArray(size_t size) throw() : NumberArray(), _value_v(NULL)
{
  resize(size);
}

CharArray::CharArray(const NumberArray& v) throw() : NumberArray(), _value_v(NULL)
{
  resize(v.size());
  for (size_t n = 0; n < size(); n++) set(n, _value_v[n]);
}

CharArray::~CharArray() throw()
{
  if (_value_v != NULL) delete [] _value_v;
}

double CharArray::get(size_t n) const throw()
{
  if (n < size()) return _value_v[n];
  else return -1;
}
void CharArray::set(size_t n, double v) throw()
{
  if (n < size()) _value_v[n] = (char)v;
}
void CharArray::resize(int s) throw()
{
  if (_value_v != NULL) delete [] _value_v;
  _value_v = new char [s];
  _size = s;
  for (size_t n = 0; n < size(); n++) set(n, 0);
}

std::string CharArray::toScript() const throw()
{
  std::stringstream ss("");
  for (size_t n = 0; n < size(); n++) {
    ss << get(n);
    if (n < size() - 1) ss << ",";
  }
  return ss.str();
}

char& CharArray::operator[](size_t n) throw()
{
  if (n < size()) return _value_v[n];
  return _value_v[0];
}

char CharArray::operator[](size_t n) const throw()
{
  if (n < size()) return _value_v[n];
  return 0;
}

void CharArray::readObject(Belle2::Reader& reader)
throw(Belle2::IOException)
{
  for (size_t n = 0; n < size(); n++) {
    _value_v[n] = reader.readChar();
  }
}

void CharArray::writeObject(Belle2::Writer& writer)
const throw(Belle2::IOException)
{
  for (size_t n = 0; n < size(); n++) {
    writer.writeChar(_value_v[n]);
  }
}

double CharArray::readElement(Belle2::Reader& reader)
throw(Belle2::IOException)
{
  return reader.readChar();
}

void CharArray::writeElement(double value, Belle2::Writer& writer)
const throw(Belle2::IOException)
{
  writer.writeChar(value);
}

File: daq/slc/extra/javalib/b2daq/src/b2daq/dqm/core/CharArray.java
package b2daq.dqm.core;

import java.io.IOException;

import b2daq.core.Reader;
import b2daq.core.Writer;

public class CharArray extends NumberArray {

    private char[] _value_v;

    public CharArray() {
        _value_v = new char[0];
        for (int n = 0; n < length(); n++) {
            _value_v[n] = 0;
        }
    }

    public CharArray(int len) {
        _value_v = new char[len];
        for (int n = 0; n < length(); n++) {
            _value_v[n] = 0;
        }
    }

    public CharArray(NumberArray array) {
        copy(array);
    }

    @Override
    public void copy(NumberArray array) {
        _value_v = new char[array.length()];
        for (int n = 0; n < length(); n++) {
            _value_v[n] = (char) array.get(n);
        }
    }

    @Override
    public void copy(char[] array) {
        _value_v = new char[array.length];
        for (int n = 0; n < length(); n++) {
            _value_v[n] = array[n];
        }
    }

    @Override
    public double get(int index) {
        if (index >= 0 && index < length()) {
            return _value_v[index];
        } else {
            return -1;
        }
    }

    public char at(int index) {
        if (index >= 0 && index < length()) {
            return _value_v[index];
        } else {
            return (char) -1;
        }
    }

    @Override
    public void set(int index, char data) {
        if (index >= 0 && index < length()) {
            _value_v[index] = data;
        }
    }

    @Override
    public void set(int index, double data) {
        if (index >= 0 && index < length()) {
            _value_v[index] = (char) data;
        }
    }

    @Override
    public int length() {
        return _value_v.length;
    }

    @Override
    public void resize(int len) {
        _value_v = new char[len];
        for (int n = 0; n < length(); n++) {
            _value_v[n] = 0;
        }
    }

    @Override
    public void writeSingle(double value, Writer writer) throws IOException {
        writer.writeChar((char) value);
    }

    @Override
    public double readSingle(Reader reader) throws IOException {
        return reader.readChar();
    }

}
